#GSOC PR : Enhance CNN architecture and training for climate data downscaling

sambhavnoobcoder opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
# Enhance CNN architecture and training for climate data downscaling

<!--- Please select appropriate Priority, Status,and Type labels-->
<!--- If you do not have permission to select labels please state which labels you would like -->

## Description
<!--- Describe your changes in detail -->
This PR introduces several improvements to our CNN model for climate data downscaling:

1. Added batch normalization layers after convolutional and dense layers
2. Implemented dropout (rate: 0.3) after batch normalization layers
3. Introduced exponential decay learning rate scheduler (initial rate: 0.001, decay steps: 1000, decay rate: 0.9)
4. Added early stopping mechanism (monitors validation loss, patience: 10 epochs)
5. Increased maximum epochs from 100 to 500

These changes aim to enhance model performance, stability, and generalization.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
These changes address key areas for improvement in our current model:
1. Reduce internal covariate shift
2. Mitigate overfitting
3. Optimize learning rate dynamics
4. Improve training efficiency
5. Increase capacity for learning complex climate patterns

By implementing these enhancements, we expect to see improved model stability, better generalization, and potentially increased prediction accuracy.

## Review Time Estimate
<!---When do you want your code reviewed by?-->
- [ ] Within one week
- [x] Immediately
- [ ] When possible

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue) <!-- please add issue number -->
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] My name is in the list of CITATION.cff
- [ ] I have updated the CHANGELOG.md.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

sambhavnoobcoder committed:
Add batch normalization to CNN model

- Inserted batch normalization layers after convolutional and dense layers
- Aims to improve model stability and performance
- May help reduce internal covariate shift and allow for higher learning rates

sambhavnoobcoder committed:
Add dropout layers to CNN model

- Inserted dropout layers after batch normalization in convolutional and dense layers
- Set dropout rate to 0.3
- Aims to reduce overfitting and improve generalization
- May enhance model robustness and performance on unseen data

sambhavnoobcoder committed:
Add exponential decay learning rate scheduler

- Implemented learning rate scheduler using exponential decay
- Initial learning rate set to 0.001
- Decay steps: 1000, decay rate: 0.9
- Aims to improve training stability and convergence
- May help fine-tune model performance over training epochs

sambhavnoobcoder committed:
Implement early stopping in CNN model

- Added early stopping callback
- Monitor: validation loss
- Patience: 10 epochs
- Restore best weights: True
- Aims to prevent overfitting and optimize training duration
- May improve model generalization and reduce unnecessary computation

sambhavnoobcoder committed:
Increase maximum number of epochs in CNN model

- Raised max epochs from 100 to 500
- Allows for potentially longer training time
- Aims to give model more opportunity to learn complex patterns
- Works in conjunction with early stopping for optimal training duration
- May lead to improved model performance and accuracy

sambhavnoobcoder committed:
Final refactor 

made some final refactoring changes to keep the code standardised and even .

mdietze commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -146,22 +146,41 @@ SDA_downscale <- function(preprocessed, date, carbon_pool, covariates, model_typ
     for (i in seq_along(carbon_data)) {
       model <- keras3::keras_model_sequential() |>
         keras3::layer_conv_1d(filters = 64, kernel_size = 1, activation = 'relu', input_shape = c(1, length(covariate_names))) |>
+        keras3::layer_batch_normalization() |>
```
Since most readers of this code will be unfamiliar with keras, I wonder if you could comment the code a bit more to explain what the different steps are doing. e.g. here you are adding batch normalization and dropout twice, and I can't tell if that's a mistake or the desired approach.

sambhavnoobcoder replied:
added comments to the code to make it more self explanatory . aimed to keep the comments concise to not cluster the code too much , but also informative enough to convey a good idea to the reader . if you feel at some place the comments need to be more descriptive , do tell me I'll make that change . the change is addressed in the commit [d2069e0](https://github.com/PecanProject/pecan/pull/3356/commits/d2069e0e36ab4b8a2604ab9629dae89dc4c06b2f)

sambhavnoobcoder replied:
Also , as for the batch normalisation and dropouts being used twice , it has been an intended behaviour on my part due to the following reason :

## First Batch Normalization:

- Purpose: Normalizes the outputs of the convolutional layer.
- Why it's used: It helps stabilize the learning process immediately after feature extraction, reducing internal covariate shift and potentially allowing for higher learning rates.

## First Dropout 

- Purpose: Randomly deactivates neurons from the convolutional layer output.
- Why it's used: It prevents overfitting on specific features extracted by the convolutional layer, encouraging the model to learn more robust and generalized representations.

## Second Batch Normalisation

- Purpose: Normalizes the outputs of the dense layer.
- Why it's needed: Even after the first normalization, the dense layer can still produce outputs with varying scales. This second normalization ensures that the inputs to the final layer are well-behaved, potentially improving the model's performance and stability in the deeper part of the network.

## Second Dropout 

- Purpose: Randomly deactivates neurons from the dense layer output.
- Why it's needed: This second dropout layer provides additional regularization at a different level of abstraction. It helps prevent co-adaptation of neurons in the dense layer, further reducing the risk of overfitting and improving generalization, especially for the higher-level features learned by the dense layer.

hence , as explained above , I've felt it to be good design decision to use them twice . In summary while the first set of batch normalization and dropout operates on the lower-level features extracted by the convolutional layer, the second set works on the higher-level features learned by the dense layer. This helps us ensure that the resultant model is more robust and gives better outputs , especailly on previously unseen data . 

I have tried to explain that a bit in the commit above without clustering the code with comments a lot as well .

sambhavnoobcoder replied:
added a few more comments , now the final commit addressing this becomes [8fdb1d5](https://github.com/PecanProject/pecan/pull/3356/commits/8fdb1d5444ba4d4ac66ea3a5f9ee58325f394a4a)

sambhavnoobcoder committed:
Added comments to architecture

added short and concise comments to the architecture in attempt to make the choice and the reasoning of the architecture design of the model self explanatory .

sambhavnoobcoder committed:
comment for CNN predictions 

added comments over the cnn prediction snippet

sambhavnoobcoder committed:
More code code comments

added more comments for clearity

sambhavnoobcoder committed:
yet more comments

this should resolve all the places comments should occur in the CNN . all the code snippets are successfully covered now .
